`mkYesodSubDispatch` doesn't seem to work with nested subsites

I'm trying to write a Yesod subsite which bundles its own static assets. The `yesod-static` package normally works well for managing static assets, but it is a subsite, and it seems nesting a subsite within another subsite doesn't work with the current `mkYesodSubDispatch` TH function.

Here is a minimal example which mostly follows the example in the [Yesod book](https://www.yesodweb.com/book/creating-a-subsite). I have also put this example together into [a repository](https://github.com/jezen/subsite-with-static).

```haskell
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE QuasiQuotes #-}
{-# LANGUAGE RankNTypes #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE ViewPatterns #-}

module Subsite.Data where

import ClassyPrelude
import Yesod
import Yesod.Static

{- HLINT ignore "Use newtype instead of data" -}
data Subsite = Subsite
  { getStatic :: Static
  }

mkYesodSubData "Subsite" [parseRoutes|
/ SubHomeR GET

-- TODO: Make this subsite work
-- /static StaticR Static getStatic
|]

newSubsite :: MonadIO m => m Subsite
newSubsite = Subsite <$> liftIO (static "static")
```

```haskell
{-# OPTIONS_GHC -fno-warn-orphans #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE QuasiQuotes #-}
{-# LANGUAGE RankNTypes #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE ViewPatterns #-}

module Subsite
  ( module Subsite.Data
  , module Subsite
  ) where

import ClassyPrelude
import Subsite.Data
import Yesod

getSubHomeR :: Yesod master => SubHandlerFor Subsite master Html
getSubHomeR = liftHandler $ defaultLayout $ do
  -- TODO: Swap this remote stylesheet for the local one
  addStylesheetRemote "https://necolas.github.io/normalize.css/8.0.1/normalize.css"
  -- TODO: Make this work
  --addStylesheet $ StaticR css_normalize_css
  [whamlet|Welcome to the subsite!|]

instance Yesod master => YesodSubDispatch Subsite master where
  yesodSubDispatch = $(mkYesodSubDispatch resourcesSubsite)
```

```haskell
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE QuasiQuotes #-}
{-# LANGUAGE RecordWildCards #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeFamilies #-}

module Main where

import ClassyPrelude hiding (Handler)
import Subsite
import Yesod
import Yesod.Core.Types

{- HLINT ignore "Use newtype instead of data" -}
data App = App
  { getSubsite :: Subsite
  }

mkMessage "App" "messages" "en"

mkYesod "App" [parseRoutes|
/ HomeR GET
/sub SubsiteR Subsite getSubsite
|]

getHomeR :: Handler Html
getHomeR = do
  now <- liftIO getCurrentTime
  defaultLayout
    [whamlet|<p>The time now is: #{tshow now}|]

instance Yesod App where

makeFoundation :: IO App
makeFoundation = App <$> newSubsite

main :: IO ()
main = makeFoundation >>= warp 3000

makeApplication :: App -> IO Application
makeApplication foundation = do
  appPlain <- toWaiAppPlain foundation
  return $ defaultMiddlewaresNoLogging appPlain

getApplicationRepl :: IO (Int, App, Application)
getApplicationRepl = do
  foundation <- makeFoundation
  app1       <- makeApplication foundation
  return (3000, foundation, app1)

shutdownApp :: App -> IO ()
shutdownApp _foundation = pure ()
```

---

When I uncomment this line in the routes:

```
 /static StaticR Static getStatic
```

…the compiler blows up like this:

```
src/Subsite.hs:28:23: error:
    • Couldn't match type: SubHandlerFor Subsite master0 content0
                     with: HandlerFor Subsite TypedContent
      Expected: Yesod.Core.Types.ParentRunner Subsite
        Actual: SubHandlerFor Subsite master0 content0
                -> Yesod.Core.Types.YesodSubRunnerEnv Subsite master0
                -> Maybe (Route Subsite)
                -> Application
    • In the first argument of ‘\ parentRunner_an3k
                                  getSub_an3l
                                  toParent_an3m
                                  env_an3n
                                  -> yesodSubDispatch
                                       Yesod.Core.Types.YesodSubRunnerEnv
                                         {Yesod.Core.Types.ysreParentRunner = parentRunner_an3k,
                                          Yesod.Core.Types.ysreGetSub = getSub_an3l,
                                          Yesod.Core.Types.ysreToParentRoute = toParent_an3m,
                                          Yesod.Core.Types.ysreParentEnv = env_an3n}’, namely
        ‘yesod-core-1.6.24.2:Yesod.Core.Class.Dispatch.subHelper’
      In the expression:
        (((((\ parentRunner_an3k getSub_an3l toParent_an3m env_an3n
               -> yesodSubDispatch
                    Yesod.Core.Types.YesodSubRunnerEnv
                      {Yesod.Core.Types.ysreParentRunner = parentRunner_an3k,
                       Yesod.Core.Types.ysreGetSub = getSub_an3l,
                       Yesod.Core.Types.ysreToParentRoute = toParent_an3m,
                       Yesod.Core.Types.ysreParentEnv = env_an3n})
              yesod-core-1.6.24.2:Yesod.Core.Class.Dispatch.subHelper)
             (\ sub_an3o -> getStatic sub_an3o))
            (\ sroute_an3p -> StaticR sroute_an3p))
           env9189_an3e)
          (((\ p_an3i r_an3j
               -> r_an3j {Network.Wai.Internal.pathInfo = p_an3i})
              restPath_an3h)
             req9189_an3f)
      In an equation for ‘helper9189_an3g’:
          helper9189_an3g ((:) "static" restPath_an3h)
            = (((((\ parentRunner_an3k getSub_an3l toParent_an3m env_an3n
                     -> yesodSubDispatch
                          Yesod.Core.Types.YesodSubRunnerEnv
                            {Yesod.Core.Types.ysreParentRunner = parentRunner_an3k,
                             Yesod.Core.Types.ysreGetSub = getSub_an3l,
                             Yesod.Core.Types.ysreToParentRoute = toParent_an3m,
                             Yesod.Core.Types.ysreParentEnv = env_an3n})
                    yesod-core-1.6.24.2:Yesod.Core.Class.Dispatch.subHelper)
                   (\ sub_an3o -> getStatic sub_an3o))
                  (\ sroute_an3p -> StaticR sroute_an3p))
                 env9189_an3e)
                (((\ p_an3i r_an3j
                     -> r_an3j {Network.Wai.Internal.pathInfo = p_an3i})
                    restPath_an3h)
                   req9189_an3f)
   |
28 |   yesodSubDispatch = $(mkYesodSubDispatch resourcesSubsite)
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Failed, one module loaded.
(0.68 secs,)
```

I _think_ this is because `YesodSubRunnerEnv` is getting the wrong value for (perhaps) `ysreParentRunner`. I also suspect this is related to https://github.com/yesodweb/yesod/issues/1002 because of @andrewthad's comment.

> I occasionally build subsites without using yesod's routing TH (to get **non-monomorphic routes**)

--- a/yesod-core/ChangeLog.md
+++ b/yesod-core/ChangeLog.md
@@ -1,5 +1,9 @@
 # ChangeLog for yesod-core
 
+## 1.6.24.3
+
+* Fix subsite-to-subsite dispatch [#1805](https://github.com/yesodweb/yesod/pull/1805)
+
 ## 1.6.24.2
 
 * No star is type [#1797](https://github.com/yesodweb/yesod/pull/1797)

--- a/yesod-core/src/Yesod/Core/Internal/TH.hs
+++ b/yesod-core/src/Yesod/Core/Internal/TH.hs
@@ -167,18 +167,10 @@ mkYesodGeneral appCxt' namestr mtys isSub f resS = do
             ]
     return (dataDec, dispatchDec)
 
-mkMDS :: (Exp -> Q Exp) -> Q Exp -> MkDispatchSettings a site b
-mkMDS f rh = MkDispatchSettings
+mkMDS :: (Exp -> Q Exp) -> Q Exp -> Q Exp -> MkDispatchSettings a site b
+mkMDS f rh sd = MkDispatchSettings
     { mdsRunHandler = rh
-    , mdsSubDispatcher =
-        [|\parentRunner getSub toParent env -> yesodSubDispatch
-                                 YesodSubRunnerEnv
-                                    { ysreParentRunner = parentRunner
-                                    , ysreGetSub = getSub
-                                    , ysreToParentRoute = toParent
-                                    , ysreParentEnv = env
-                                    }
-                              |]
+    , mdsSubDispatcher = sd
     , mdsGetPathInfo = [|W.pathInfo|]
     , mdsSetPathInfo = [|\p r -> r { W.pathInfo = p }|]
     , mdsMethod = [|W.requestMethod|]
@@ -199,7 +191,20 @@ mkDispatchInstance :: Type                      -- ^ The master site type
                    -> [ResourceTree c]          -- ^ The resource
                    -> DecsQ
 mkDispatchInstance master cxt f res = do
-    clause' <- mkDispatchClause (mkMDS f [|yesodRunner|]) res
+    clause' <- 
+        mkDispatchClause 
+            (mkMDS 
+                f 
+                [|yesodRunner|] 
+                [|\parentRunner getSub toParent env -> yesodSubDispatch
+                    YesodSubRunnerEnv
+                    { ysreParentRunner = parentRunner
+                    , ysreGetSub = getSub
+                    , ysreToParentRoute = toParent
+                    , ysreParentEnv = env
+                    }
+                |])
+            res
     let thisDispatch = FunD 'yesodDispatch [clause']
     return [instanceD cxt yDispatch [thisDispatch]]
   where
@@ -207,7 +212,20 @@ mkDispatchInstance master cxt f res = do
 
 mkYesodSubDispatch :: [ResourceTree a] -> Q Exp
 mkYesodSubDispatch res = do
-    clause' <- mkDispatchClause (mkMDS return [|subHelper|]) res
+    clause' <- 
+        mkDispatchClause
+            (mkMDS 
+                return 
+                [|subHelper|] 
+                [|\_ getSub toParent env -> yesodSubDispatch
+                    YesodSubRunnerEnv
+                    { ysreParentRunner = ysreParentRunner env
+                    , ysreGetSub = getSub . ysreGetSub env
+                    , ysreToParentRoute = ysreToParentRoute env . toParent
+                    , ysreParentEnv = ysreParentEnv env
+                    }
+                |]) 
+        res
     inner <- newName "inner"
     let innerFun = FunD inner [clause']
     helper <- newName "helper"

--- a/yesod-core/src/Yesod/Core/Internal/TH.hs
+++ b/yesod-core/src/Yesod/Core/Internal/TH.hs
@@ -4,6 +4,7 @@
 {-# LANGUAGE FlexibleInstances #-}
 {-# LANGUAGE MultiParamTypeClasses #-}
 {-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE RankNTypes #-}
 module Yesod.Core.Internal.TH where
 
 import Prelude hiding (exp)
@@ -22,6 +23,7 @@ import Text.ParserCombinators.Parsec.Char (alphaNum, spaces, string, char)
 
 import Yesod.Routes.TH
 import Yesod.Routes.Parse
+import Yesod.Core.Content (ToTypedContent (..))
 import Yesod.Core.Types
 import Yesod.Core.Class.Dispatch
 import Yesod.Core.Internal.Run
@@ -217,14 +219,7 @@ mkYesodSubDispatch res = do
             (mkMDS 
                 return 
                 [|subHelper|] 
-                [|\_ getSub toParent env -> yesodSubDispatch
-                    YesodSubRunnerEnv
-                    { ysreParentRunner = ysreParentRunner env
-                    , ysreGetSub = getSub . ysreGetSub env
-                    , ysreToParentRoute = ysreToParentRoute env . toParent
-                    , ysreParentEnv = ysreParentEnv env
-                    }
-                |]) 
+                [|subTopDispatch|]) 
         res
     inner <- newName "inner"
     let innerFun = FunD inner [clause']
@@ -236,6 +231,26 @@ mkYesodSubDispatch res = do
                     [innerFun]
                 ]
     return $ LetE [fun] (VarE helper)
+    
+subTopDispatch :: 
+    (YesodSubDispatch sub master) =>
+        (forall content. ToTypedContent content =>
+            SubHandlerFor child master content ->
+            YesodSubRunnerEnv child master ->
+            Maybe (Route child) ->
+            W.Application
+        ) ->
+        (mid -> sub) ->
+        (Route sub -> Route mid) ->
+        YesodSubRunnerEnv mid master ->
+        W.Application
+subTopDispatch _ getSub toParent env = yesodSubDispatch
+            (YesodSubRunnerEnv
+            { ysreParentRunner = ysreParentRunner env
+            , ysreGetSub = getSub . ysreGetSub env
+            , ysreToParentRoute = ysreToParentRoute env . toParent
+            , ysreParentEnv = ysreParentEnv env
+            })
 
 instanceD :: Cxt -> Type -> [Dec] -> Dec
 instanceD = InstanceD Nothing

--- a/yesod-core/test/YesodCoreTest.hs
+++ b/yesod-core/test/YesodCoreTest.hs
@@ -9,6 +9,7 @@ import YesodCoreTest.Meta
 import YesodCoreTest.Links
 import YesodCoreTest.Header
 import YesodCoreTest.NoOverloadedStrings
+import YesodCoreTest.SubSub
 import YesodCoreTest.InternalRequest
 import YesodCoreTest.ErrorHandling
 import YesodCoreTest.Cache
@@ -43,6 +44,7 @@ specs = do
       mediaTest
       linksTest
       noOverloadedTest
+      subSubTest
       internalRequestTest
       errorHandlingTest
       cacheTest

--- a/yesod-core/test/YesodCoreTest/SubSub.hs
+++ b/yesod-core/test/YesodCoreTest/SubSub.hs
@@ -1,0 +1,50 @@
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module YesodCoreTest.SubSub where
+
+import Test.Hspec
+
+import Yesod.Core
+import Network.Wai.Test
+import qualified Data.Text as T
+import qualified Data.ByteString.Lazy.Char8 as L8
+
+import YesodCoreTest.SubSubData
+
+data App = App { getOuter :: OuterSubSite }
+
+instance Yesod App
+
+getSubR :: SubHandlerFor InnerSubSite master T.Text
+getSubR = return $ T.pack "sub"
+
+instance YesodSubDispatch OuterSubSite master where
+  yesodSubDispatch = $(mkYesodSubDispatch resourcesOuterSubSite)
+
+instance YesodSubDispatch InnerSubSite master where
+  yesodSubDispatch = $(mkYesodSubDispatch resourcesInnerSubSite)
+
+mkYesod "App" [parseRoutes|
+/ OuterSubSiteR OuterSubSite getOuter
+|]
+
+app :: App
+app = App { getOuter = OuterSubSite { getInner = InnerSubSite }}
+
+runner :: Session () -> IO ()
+runner f = toWaiApp app >>= runSession f
+
+case_subSubsite :: IO ()
+case_subSubsite = runner $ do
+  res <- request defaultRequest
+  assertBody (L8.pack "sub") res
+  assertStatus 200 res
+
+subSubTest :: Spec
+subSubTest = describe "YesodCoreTest.SubSub" $ do
+  it "sub_subsite" case_subSubsite

--- a/yesod-core/test/YesodCoreTest/SubSubData.hs
+++ b/yesod-core/test/YesodCoreTest/SubSubData.hs
@@ -1,0 +1,20 @@
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE TypeFamilies #-}
+
+module YesodCoreTest.SubSubData where
+
+import Yesod.Core
+
+
+data OuterSubSite = OuterSubSite { getInner :: InnerSubSite }
+
+data InnerSubSite = InnerSubSite
+
+mkYesodSubData "InnerSubSite" [parseRoutes|
+/ SubR GET
+|]
+
+mkYesodSubData "OuterSubSite" [parseRoutes|
+/ InnerSubSiteR InnerSubSite getInner
+|]

--- a/yesod-core/yesod-core.cabal
+++ b/yesod-core/yesod-core.cabal
@@ -1,5 +1,5 @@
 name:            yesod-core
-version:         1.6.24.2
+version:         1.6.24.3
 license:         MIT
 license-file:    LICENSE
 author:          Michael Snoyman <michael@snoyman.com>

--- a/yesod-core/yesod-core.cabal
+++ b/yesod-core/yesod-core.cabal
@@ -173,6 +173,8 @@ test-suite tests
                    YesodCoreTest.StubSslOnly
                    YesodCoreTest.StubStrictSameSite
                    YesodCoreTest.StubUnsecured
+                   YesodCoreTest.SubSub
+                   YesodCoreTest.SubSubData
                    YesodCoreTest.WaiSubsite
                    YesodCoreTest.Widget
                    YesodCoreTest.YesodTest

